src/app/ReadClient.cpp

```diff
@@ -77,95 +78,112 @@ void ReadClient::MoveToState(const ClientState aTargetState)
+                .ClusterId(apAttributePathParamsList[index].mClusterId);
+
+            AttributePathSelector * selector = apAttributePathParamsList[index].mpSelector;
+
+            while (selector->mpNext != nullptr)
             {
-                AttributePath::Builder attributePathBuilder = attributePathListBuilder.CreateAttributePathBuilder();
-                attributePathBuilder.NodeId(apAttributePathParamsList[index].mNodeId)
-                    .EndpointId(apAttributePathParamsList[index].mEndpointId)
-                    .ClusterId(apAttributePathParamsList[index].mClusterId);
-                if (apAttributePathParamsList[index].mFlags == AttributePathFlags::kFieldIdValid)
+                if (selector->mFlag == AttributePathSelectorFlag::kFieldIdValid)
                 {
-                    attributePathBuilder.FieldId(apAttributePathParamsList[index].mFieldId);
+                    attributePathBuilder.FieldId(selector->mFieldId);
                 }
-                else if (apAttributePathParamsList[index].mFlags == AttributePathFlags::kListIndexValid)
+                else if (selector->mFlag == AttributePathSelectorFlag::kListIndexValid)
                 {
-                    attributePathBuilder.ListIndex(apAttributePathParamsList[index].mListIndex);
+                    attributePathBuilder.ListIndex(selector->mListIndex);
                 }
                 else
                 {
-                    err = CHIP_ERROR_INVALID_ARGUMENT;
-                    ExitNow();
+                    ExitNow(err = CHIP_ERROR_INVALID_ARGUMENT);
                 }
-                attributePathBuilder.EndOfAttributePath();
-                SuccessOrExit(err = attributePathBuilder.GetError());
+                selector = selector->mpNext;
             }
-            attributePathListBuilder.EndOfAttributePathList();
-            SuccessOrExit(err = attributePathListBuilder.GetError());
+
+            attributePathBuilder.EndOfAttributePath();
+            SuccessOrExit(err = attributePathBuilder.GetError());
         }
+        attributePathListBuilder.EndOfAttributePathList();
+        SuccessOrExit(err = attributePathListBuilder.GetError());
+    }
 
-        request.EndOfReadRequest();
-        SuccessOrExit(err = request.GetError());
+    request.EndOfReadRequest();
+    SuccessOrExit(err = request.GetError());
 
-        err = writer.Finalize(&msgBuf);
-        SuccessOrExit(err);
-    }
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdminId, EventPathParams * apEventPathParamsList,
+                                       size_t aEventPathParamsListSize, AttributePathParams * apAttributePathParamsList,
+                                       size_t aAttributePathParamsListSize, EventNumber aEventNumber)
+{
+    // TODO: SendRequest parameter is too long, need to have the structure to represent it
```

Comment:
It looks like at least the pattern lf `Foo *foo, size_t fooLength` could be replaced with some sort of Span.